### PR TITLE
Add state guide

### PR DESCRIPTION
**Pull request by Lunarmagpie**
Im doing this as a draft because editing is hard to explain without good methods to help with that.

### CODE REVIEW

File: docs/source/guides/state.md [hunk 1]
```diff
@@ -1 +1,135 @@
...
+is to create all of them again.
+
+```python
+@flare.button(label="Click Me!")
+async def button(
+    ctx: flare.Context,
+    number: int,
+) -> None: 
+    await ctx.edit_response(
+        component=flare.Row(button(number=number+1))
```
Comment (HyperGH) [inline docs/source/guides/state.md:62]: is this syntax still valid?

File: docs/source/guides/state.md [hunk 2]
```diff
@@ -1 +1,135 @@
...
+
+    # Edit the values for `counter_button` in the list of
+    # rows.
+    me = rows[0][0]
+    me.number += 1
+
+    await ctx.edit_response(
+        content=number,
+        # Build all the rows.
+        components=await asyncio.gather(*rows),
```
Comment (HyperGH) [inline docs/source/guides/state.md:105]: or is it this?

File: docs/source/guides/state.md [hunk 3]
```diff
@@ -1 +1,135 @@
...
+        print(self.b)
+
+        # Type safe component attribute editing
+        self.a = 5
+        self.b = 10
+
+
+        await ctx.edit_response(
+            content=number,
+            component=await flare.Row(self),
```
Comment (HyperGH) [inline docs/source/guides/state.md:133]: or maybe this?

**Review by HyperGH — CHANGES_REQUESTED** [review]
Inconsistent handling of how components are passed around, and at least one case of using possibly old dead code. I'd recommend testing if these examples still work.

File: docs/source/guides/state.md (hunk 2 above)
Comment (Lunarmagpie) [reply docs/source/guides/state.md:105]: I should document why gather is needed here better.

File: docs/source/guides/state.md (hunk 3 above)
Comment (Lunarmagpie) [reply docs/source/guides/state.md:133]: Its this

**Review by HyperGH — APPROVED** [review]

**Pull request merged**